modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_remote_downsample_strategy.py

```diff
@@ -5,39 +5,49 @@
 
 
 def get_tensors_subset(
-    indexes: torch.Tensor, data: Union[torch.Tensor, dict], target: torch.Tensor, sample_ids: list
-) -> tuple[Union[torch.Tensor, dict], torch.Tensor, list]:
+    selected_indexes: list[int], data: Union[torch.Tensor, dict], target: torch.Tensor, sample_ids: list
+) -> tuple[Union[torch.Tensor, dict], torch.Tensor]:
+    # Starting from the batch, we want to keep only the selected samples
+
+    # first of all we compute the position of each selected index within the batch
```

Comment:
This is the index map from the other comment, right? It's a bit different now, right? You had a great explanation in your last Github comment, you might want to adapt and copy it here because you've written it anyways already and it was very helpful in understanding what is happening

Reply:
More or less, the role of this function specifically has changed since `index_sampleid_map` was introduced.

Before, when the downsampler returned the indices, this function simply had to take the samples of the various tensors (so if the selected indices were 1 and 3 it would take samples 1 and 3, labels 1 and 3 and so on). This function is only needed for batch-then-sample (where we have the samples and labels already available).

Now, the downsampler returns the sample ids. This way, we have to do the reverse process (i.e. work out which index the various sample_ids correspond to) and then repeat the same operation. So, for example, we now get that the selected ids are 132 and 154 and that all the ids are [102, 132, 15, 154, 188]. As a result, we get that the corresponding ids are 1 and 3 and then it is exactly the same as before.

Moving index_sampleid_map within the downsampler simplifies STB but slightly complicates BTS (this extra operation is necessary). However, I think it is an acceptable price to pay for having a single interface.

Reply:
Btw I can add the comment where index_sampleid_map is actually used (and it's exactly the same as the comment)

Reply:
Comment added when `self.index_sampleid_map` is created

Reply:
Ahh ok makes sense - thanks for the explanation!